### PR TITLE
Allow parallelize_module to get device_mesh from ambient context

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134247

This PR is for supporting calling `parallelize_module` from within a model definition, making the model a parallel one.

Calling `parallelize_module` is an alternative to maintaining a set of `ColumnWiseLinear`, `RowWiseLinear`, etc, while still being able to directly author a parallel model. 

(The motivation for authoring a parallel model is that there may be other distributed operations, which may not be easily captured by any module, see the forward function below. Alternatively speaking, the purpose is to exploit the expressiveness of DTensor -- we need to first create DTensors before calling ops on them. Having parallelized modules in model is one way of creating DTensors.)

For example:
```
class FeedForward(nn.Module):
    def __init__(self, config: TransformerArgs) -> None:
        super().__init__()
        w1 = nn.Linear(config.dim, config.hidden_dim, bias=False)
        w2 = nn.Linear(config.hidden_dim, config.dim, bias=False)
        w3 = nn.Linear(config.dim, config.hidden_dim, bias=False)
        self.w1 = parallelize_module(w1, Colwise)
        self.w2 = parallelize_module(w2, Rowwise)
        self.w3 = parallelize_module(w3, Colwise)

    def forward(self, x: Tensor) -> Tensor:
        y: DTensor = self.w2(F.silu(self.w1(x)) * self.w3(x))
        # y is a DTensor with Partial placement; we can return it as is.
        return y
        # Or we can convert it to Replicate -- there is modeling flexibility here.
        return y.redistribute(Replicate())


with device_mesh:
    model = FeedForward(config)
    # Now model is a model parallelized onto device_mesh

y = model(x)

```

The `device_mesh` actually used for `parallelize_module` would be retrieved from the ambient context.

Calling `parallelize_module` from within model hierarchy also saves the use of *FQNs* as in the out-of-model annotation case.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/tensor/parallel/api.py [hunk 1]
```diff
@@ -21,8 +22,8 @@
 
 def parallelize_module(  # type: ignore[return]
     module: nn.Module,
-    device_mesh: DeviceMesh,
-    parallelize_plan: Union[ParallelStyle, Dict[str, ParallelStyle]],
+    device_mesh: Optional[DeviceMesh] = None,
+    parallelize_plan: Optional[Union[ParallelStyle, Dict[str, ParallelStyle]]] = None,
```
Comment (awgu) [inline torch/distributed/tensor/parallel/api.py:26]: maybe explain why you want this?
why not just skip calling `parallelize_module` at all?

Comment (kwen2501) [reply torch/distributed/tensor/parallel/api.py:26]: You are so fast in reviewing it, even before I finish writing the PR desc. 
Now the motivation is added :)

**Comment by fduwjj** [comment]
Per your example, how do you annotate the input for your nn.module? So you will call `parallelize_module(FeedForward, input_plan)` something like that? So aside from avoiding specifying long FQN list, what other benefits are there?

**Comment by kwen2501** [comment]
@fduwjj The manipulation to inputs can be written in the forward function (just like how we manipulate the output above.) In fact, we can manipulate activations too.

```
    def forward(self, x: Tensor) -> Tensor:
        # Manipulate inputs
        x = x.redistribute(...)
        y: DTensor = self.w2(F.silu(self.w1(x)) * self.w3(x))
        ....
```

**Comment by fduwjj** [comment]
Another question I have is, why can't users directly write their module using DTensor? because we don't have a nn.Linear which is directly built on DTensor, so we need this swap?

Since you already do lots of DTensor operations. For the purpose of flexibility, why can't users do:
```
self.w1 = nn.Linear(...)
self.w1.weight = nn.Parameter(DTensor_shard(self.w1.weight))
```

I mean I don't have answer here, but I am just thinking what is most flexible way for users. And this way, users just swap whatever part to DTensor but they need to take care not to mix Tensor and DTensor together.

**Comment by kwen2501** [comment]
```
self.w1 = nn.Linear(...)
self.w1.weight = nn.Parameter(DTensor_shard(self.w1.weight))
```
They can do this, just not that modular, as compared to calling `parallelize_module` or using `ColumnWiseLinear` class.
The assumption is that authors still want some abstraction than writing pure DTensor programs.

**Review by wanchaol — COMMENTED** [review]
I think the change looks reasonable to me. Please update documentations to reflect these fields can be optional (and the behavior when they are optional), and add tests

File: torch/distributed/tensor/parallel/api.py [hunk 2]
```diff
@@ -67,8 +68,15 @@ def parallelize_module(  # type: ignore[return]
     """
     torch._C._log_api_usage_once("torch.distributed.tensor.parallel.parallelize_module")
 
+    device_mesh = device_mesh or _mesh_resources.get_current_mesh()
     _validate_tp_mesh_dim(device_mesh)
 
+    if parallelize_plan is None:
+        warnings.warn(
+            "No parallelize_plan is provided, so parallelize_module does nothing."
```
Comment (wz337) [inline torch/distributed/tensor/parallel/api.py:76]: minor: maybe add `does nothing to module name` since you are likely to have multiple modules.

**Review by tianyu-l — COMMENTED** [review]
I didn't fully appreciate the change, in that:
In your example, since we know `class FeedForward(nn.Module)` will be init in a distributed way, and needs to be called under some device_mesh context, why can't we put a DeviceMesh into the init args? i.e.
```
class FeedForward(nn.Module):
    def __init__(self, config: TransformerArgs, mesh: DeviceMesh) -> None:
        super().__init__()
        w1 = nn.Linear(config.dim, config.hidden_dim, bias=False)
        w2 = nn.Linear(config.hidden_dim, config.dim, bias=False)
        w3 = nn.Linear(config.dim, config.hidden_dim, bias=False)
        self.w1 = parallelize_module(w1, mesh, Colwise)
        self.w2 = parallelize_module(w2, mesh, Rowwise)
        self.w3 = parallelize_module(w3, mesh, Colwise)
```

**Comment by kwen2501** [comment]
@tianyu-l Good question. I guess there are a couple reasons:
(1) for the model's creation signature to remain the same as non-parallel case.
(2) for the model's code to look less heavy.
(3) avoid having to pass mesh from parent module to child modules.

File: torch/distributed/tensor/parallel/api.py [hunk 3]
```diff
@@ -67,8 +68,15 @@ def parallelize_module(  # type: ignore[return]
     """
     torch._C._log_api_usage_once("torch.distributed.tensor.parallel.parallelize_module")
 
+    device_mesh = device_mesh or _mesh_resources.get_current_mesh()
```
Comment (tianyu-l) [inline torch/distributed/tensor/parallel/api.py:71]: If we have nD DeviceMesh, what this line of code will retrieve? Would it intelligently fetch the TP mesh?

**Review by tianyu-l — COMMENTED** [review]
> I guess there are a couple reasons:
(1) for the model's creation signature to remain the same as non-parallel case.
(2) for the model's code to look less heavy.
(3) avoid having to pass mesh from parent module to child modules.

Re (1): I don't know if this is pro or con -- having the mesh in signature could help user be aware they are using a non-parallel module.
Re (2) and (3): I agree

Let's also consider potential downside:
(a) `parallelize_plan` should be required, but now we are sacrificing this to allow for a more flexible `device_mesh`. To me this is a trade-off, rathe than pure win.
(b) inline commented: do we always get the right device_mesh from context, when there are n-dimensional root DeviceMesh?
(c) this is a one-way move: after adopting this change, going back (for other reasons we might not be able to forsee) would be BC-breaking.

I'm a bit concerned about (a) and (b). I wonder what you and other people think.

File: torch/distributed/tensor/parallel/api.py (hunk 3 above)
Comment (kwen2501) [reply torch/distributed/tensor/parallel/api.py:71]: Whether the code is:
```
parallelize_module(mod, mesh, plan)
```
or 
```
with mesh:
    parallelize_module(mod, plan)
```
the user has responsibility to make sure `mesh` is a correct, intended one.

Comment (kwen2501) [reply torch/distributed/tensor/parallel/api.py:71]: If `mesh` is wrong, the above two pieces of code would also fail in the same way.

**Comment by kwen2501** [comment]
Re (a):
`mesh` is still in fact required. If the result of 
`device_mesh or _mesh_resources.get_current_mesh()` is None, I think an error will be returned: "No active device mesh is found."

Also want to note that APIs like `distribute_tensor` or `distribute_module` also support such usage (which to me seems like a helpful feature).

File: torch/distributed/tensor/parallel/api.py (hunk 3 above)
Comment (tianyu-l) [reply torch/distributed/tensor/parallel/api.py:71]: this makes sense to me.

but would this code run? I thought it at least should be:
`parallelize_module(mod, parallelize_plan=plan)` which doesn't make it that simple compared with `parallelize_module(mod, mesh, plan)`

**Comment by tianyu-l** [comment]
> Re (a): `mesh` is still in fact required. If the result of `device_mesh or _mesh_resources.get_current_mesh()` is None, I think an error will be returned: "No active device mesh is found."

When I say required, I mean required by the method signature.

> Also want to note that APIs like `distribute_tensor` or `distribute_module` also support such usage (which to me seems like a helpful feature).

They look different to me. For `distribute_tensor` and `distribute_module`, each optional arg has meaningful default behavior. But for `parallelize_plan`, it in nature is not optional. Here we are only marking it optional because we want to mark another previous arg optional (`device_mesh`), which is making shortcuts to bypass Python constraints.

I'd be less concerned if the order of `device_mesh` and `parallelize_plan` is switched and only `device_mesh` is marked optional.

**Comment by kwen2501** [comment]
Yeah, I agree that it would have been nicer if the order of `device_mesh` and `parallelize_plan` is switched. 
In the PR, I supported `parallelize_plan=None` by saying this is a no-op. Would that lessen your concern?

**Comment by kwen2501** [comment]
@tianyu-l This may seem to be the biggest motivation behind this change:
a separation between distributed runtime and model layout style.

`device_mesh` represents a runtime, in that it would have been embodied to concrete devices, and a communicator among them.

`parallelize_plan` represents how data of a model will be sharded, column wise or row wise, etc. It is a style of data layout that does not necessarily have to bind to a concrete device mesh at the time of description.

**Comment by tianyu-l** [comment]
> In the PR, I supported `parallelize_plan=None` by saying this is a no-op.

`parallelize_plan` is the core part of this method, and cannot be `None` -- calling the function should fail without a `parallelize_plan`.

I'm OK with setting `device_mesh = None` and get it from ambient context -- I think that's good change. Do you think we can mark `device_mesh: DeviceMesh | None` for this purpose? (yeah I hope we could switch the order of these two)

**Comment by kwen2501** [comment]
@wanchaol Sorry to bother you. 
@tianyu-l would like a sign-off on the API signature change.
Could you please sign it off (or reject it)? Thanks!

**Comment by wanchaol** [comment]
I still think the API signature change sounds good to me. Please add tests and update documentation to reflect the signature change

> parallelize_plan is the core part of this method, and cannot be None -- calling the function should fail without a parallelize_plan.

@tianyu-l I think it's ok for this to be default to None and throw a warning to the user when it's the case. The API intention is to parallelize a module with a plan/style, if user intentionally don't pass a plan then we could return the original (untouched) module, which looks a reasonable behavior to me.

>  I hope we could switch the order of these two

Unfortunately we can't switch the argument order (that considered as a BC breaking change) in python. I feel the `device_mesh` being the second argument also matches the `distribute_tensor/module` signature. The downside here is mainly it forced the `parallelize_plan` to be optional, which I feel we have a reasonable default here as mentioned above.

**Comment by tianyu-l** [comment]
Cool, then I'm ok with it.
@kwen2501 if you make tests and documentation ready, I can help review again.

**Review by tianyu-l — APPROVED** [review]
lgtm. Can you apply @wz337's suggestion in https://github.com/pytorch/pytorch/pull/134247#discussion_r1728054745 before merge?

**Comment by kwen2501** [comment]
Due to the design of the API, it is not always possible to get a module name (e.g. in the use case proposed).

**Pull request closed without merging**